vault/activity_log_test.go

```diff
@@ -5084,3 +5086,193 @@ func TestActivityLog_reportPrecomputedQueryMetrics(t *testing.T) {
+		testname := fmt.Sprintf("%s", i)
+		t.Run(testname, func(t *testing.T) {
+			// Normalize to start of month to prevent end of month weirdness
+			startOfMonth := timeutil.StartOfMonth(time.Now().UTC())
+
+			storage := &logical.InmemStorage{}
+			coreConfig := &CoreConfig{
+				CredentialBackends: map[string]logical.Factory{
+					"userpass": userpass.Factory,
+				},
+				Physical: storage.Underlying(),
+			}
+
+			cluster := NewTestCluster(t, coreConfig, nil)
+			cluster.Start()
+			defer cluster.Cleanup()
+			core := cluster.Cores[0].Core
+			TestWaitActive(t, core)
+
+			a := core.activityLog
+			ctx := namespace.RootContext(nil)
+			var err error
+
+			namespaces := make([]*namespace.Namespace, 0, 6)
+			mounts := make(map[string][]*MountEntry)
+			ns := namespace.RootNamespace
+			namespaces = append(namespaces, ns)
+
+			// Add two userpass mounts to the root namespace
+			for i := 0; i < 1; i++ {
+				me1 := &MountEntry{
+					Table: credentialTableType,
+					Path:  "up1/",
+					Type:  "userpass",
+				}
+				err = core.enableCredential(namespace.ContextWithNamespace(ctx, namespaces[i]), me1)
+				require.NoError(t, err)
+
+				me2 := &MountEntry{
+					Table: credentialTableType,
+					Path:  "up2/",
+					Type:  "userpass",
+				}
+				err = core.enableCredential(namespace.ContextWithNamespace(ctx, namespaces[i]), me2)
+				require.NoError(t, err)
+				mounts[namespaces[i].ID] = []*MountEntry{me1, me2}
+			}
+
+			// Generate data for two months ago
+			clientPrefix := 1
+			var entityRecordsMonth2 []*activity.EntityRecord
+			for namespaceId, mountSlice := range mounts {
+				for mountIndex, mount := range mountSlice {
+					entityRecordsMonth2 = append(entityRecordsMonth2, generateClientData(0, tt.twoMonthsAgo[0][mountIndex], mount.Accessor, namespaceId, fmt.Sprintf("%d", clientPrefix))...)
+				}
+			}
+
+			// Generate data for a month ago
+			clientPrefix += 1
+			var entityRecordsMonth1 []*activity.EntityRecord
+			for namespaceId, mountSlice := range mounts {
+				for mountIndex, mount := range mountSlice {
+					entityRecordsMonth2 = append(entityRecordsMonth1, generateClientData(0, tt.oneMonthAgo[0][mountIndex], mount.Accessor, namespaceId, fmt.Sprintf("%d", clientPrefix))...)
+				}
+			}
+
+			startOfTwoMonthsAgo := timeutil.StartOfMonth(startOfMonth.AddDate(0, -2, 0)).UTC()
+			startOfOneMonthAgo := timeutil.StartOfMonth(startOfMonth.AddDate(0, -1, 0)).UTC()
+			generatePreviousMonthClientData(t, core, startOfTwoMonthsAgo, entityRecordsMonth2)
+			generatePreviousMonthClientData(t, core, startOfOneMonthAgo, entityRecordsMonth1)
+
+			// Generate the current months data
+			clientPrefix += 1
+			var currentMonthData []*activity.EntityRecord
+			for namespaceId, mountSlice := range mounts {
+				for mountIndex, mount := range mountSlice {
+					currentMonthData = append(currentMonthData, generateClientData(0, tt.currentMonth[0][mountIndex], mount.Accessor, namespaceId, fmt.Sprintf("%d", clientPrefix))...)
```

Comment:
looks like generateClientData and generatePreviousMonthClientData are both defined in an enterprise file. Can we move them to CE?

Reply:
Oops! Moved these over!

Reply:
refactored this to use clientcountutil